storage/gcloud/aio/storage/storage.py

```diff
@@ -115,9 +126,113 @@ def __init__(self, project, service_file, token=None, session=None):
+                            retries=5, retry_sleep=1.):
+            tries = 0
+            while True:
+                resp = await session.put(session_URI, headers=headers, data=data, timeout=timeout)
+                tries += 1
+                if resp.status == 200:
+                    break
+                else:
+                    headers.update({'Content-Range': '*/*'})
+                    asyncio.sleep(retry_sleep)
+                if tries > retries:
+                    resp.raise_for_status()
+                    break
+
+            upload_response = await resp.json()
+
+            return upload_response
+
+        session_URI = await initiate_upload_session(object_name, headers, session)
+        upload_response = await do_upload(session_URI, data, timeout, headers, session)
+
+        return upload_response
+
+    async def _download(self, bucket, object_name, params=None, session=None):
+
+        def get_content_type_and_encoding(response_header: dict):
+            content_type_and_encoding = response_header.get('Content-Type')
+            content_type_and_encoding_split = content_type_and_encoding.split(';')
+            content_type = content_type_and_encoding_split[0].lower().strip()
+
+            encoding = None
+            if len(content_type_and_encoding_split) > 1:
+                encoding_str = content_type_and_encoding_split[1].lower().strip()
+                encoding = encoding_str.split('=')[-1]
+
+            return (content_type, encoding)
+
+        token = await self.token.get()
+        # https://cloud.google.com/storage/docs/json_api/#encoding
+        encoded_object_name = quote(object_name, safe='')
+        url = f'{STORAGE_API_ROOT}/{bucket}/o/{encoded_object_name}'
+        headers = {
+            'Authorization': f'Bearer {token}',
+        }
+
+        if not self.session:
+            self.session = aiohttp.ClientSession(conn_timeout=10,
+                                                 read_timeout=10)
+        session = session or self.session
+        response = await session.get(url, headers=headers, params=params or {},
+                                     timeout=60)
+        response.raise_for_status()
+        response_header = dict(response.headers)
+
+        content_type, encoding = get_content_type_and_encoding(response_header)
+
+        if content_type == 'application/octet-stream':
+            result = await response.read()
+        elif content_type in ('text/plain', 'text/html', 'application/json'):
+            result = await response.text(encoding)
+        else:
+            result = await response.read()
+
+        return result
+
+    def preprocess_data_(self, in_data):
+        if in_data is None:
+            out_data = ''
+        elif isinstance(in_data, str) or isinstance(in_data, bytes):
+            out_data = in_data
+        else:
+            raise RuntimeError('Currently, str and bytes -- exiting')
```

Comment:
A more specific error would be great here. Maybe `raise TypeError(f'unsupported upload type {type(in_data)}')`